### PR TITLE
Procedural macros for creating constant values from strings

### PR DESCRIPTION
**Pull request by rasmus-kirk**
## Purpose

Addresses #385 

## Changes

Added macros for creating constant values from strings for the following types:
  - `AccountAddress`
  - `ContractAddress`
  - `ModuleReference`
  - `PublicKeyEcdsaSecp256k1`
  - `PublicKeyEd25519`
  - `SignatureEcdsaSecp256k1`
  - `SignatureEd25519`

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-std-derive/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+rust-version = "1.65"
+license = "MPL-2.0"
+description = "Procedural macros to generate concordium type constants from strings, that makes writing smart contracts on the Concordium blockchain easier."
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+[lib]
+proc-macro = true
+
+[dependencies]
+# Replace with new version, obviously...
```
Comment (rasmus-kirk) [inline concordium-std-derive/Cargo.toml:15]: Todo

File: concordium-std-derive/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+};
+use proc_macro::TokenStream;
+use std::str::FromStr;
+use syn::{parse_macro_input, LitStr};
+
+/// Procedural macro for instantiating account addresses.
+/// Input must be a valid base58-encoding.
+#[proc_macro]
+pub fn account_address(item: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(item as LitStr);
```
Comment (rasmus-kirk) [inline concordium-std-derive/src/lib.rs:15]: This also correctly handles the case where we pass in anything that is _not_ a string literal, fx:

```rust
const ACC: AccountAddress = account_address!(0);
```

Will correctly be caught, and produce the following error:

```
error: Invalid account address: Invalid Base58Check encoding.
 --> tests/test-programs/fail-acc-invalid-base58.rs:6:46
  |
6 | const ACC: AccountAddress = account_address!("30Bx2h5Y2veb4hZgAJWPrr8RyQESKm5TjzF3ti1QQ4VSYLwK1G");
  |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: concordium-std-derive/Cargo.toml (hunk 1 above)
Comment (abizjak) [reply concordium-std-derive/Cargo.toml:15]: You don't have to use the `git` dependency. The contracts-common is part of concordium-base, which is a submodule of this repository. Please use taht dependency.

File: concordium-std-derive/src/lib.rs (hunk 2 above)
Comment (abizjak) [reply concordium-std-derive/src/lib.rs:15]: That does not look like the right error. We want to tell the user they need a string literal.

File: concordium-std-derive/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        }
+    };
+
+    match format!("AccountAddress({:?})", addr_bytes).parse() {
+        Ok(o) => o,
+        Err(e) => {
+            let err = syn::Error::new(input.span(), format!("LexError: {}", e));
+            err.to_compile_error().into()
+        }
+    }
```
Comment (abizjak) [inline concordium-std-derive/src/lib.rs:32]: So we can't use `quote!` to construct this? Something like
```
quote!(AccountAddress(#addr_bytes)
```

I don't really like parsing this thing again.

File: concordium-std-derive/tests/test-programs/fail-acc-multiple-arguments.stderr [hunk 4]
```diff
@@ -0,0 +1,5 @@
+error: unexpected token
+ --> tests/test-programs/fail-acc-multiple-arguments.rs:6:98
+  |
+6 | const ACC: AccountAddress = account_address!("3kBx2h5Y2veb4hZgAJWPrr8RyQESKm5TjzF3ti1QQ4VSYLwK1G", "3kBx2h5Y2veb4hZgAJWPrr8RyQESKm5TjzF3t...
+  |                                                                                                  ^
```
Comment (abizjak) [inline concordium-std-derive/tests/test-programs/fail-acc-multiple-arguments.stderr:5]: That's not the best error we can have. Could we have this case print also 'expected a string literal'?

File: concordium-std-derive/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+/// Procedural macro for instantiating contract addresses.
+/// Input must be of the form "<index,subindex>", where index and subindex
+/// are integers.
```
Comment (abizjak) [inline concordium-std-derive/src/lib.rs:137]:
```suggestion
/// are unsigned integers.
```

**Comment by abizjak** [comment]
Could you also please go over the example contracts and use these macros in tests there.

**Comment by rasmus-kirk** [comment]
Yes, of course

File: concordium-std/src/impls.rs [hunk 6]
```diff
@@ -765,6 +765,7 @@ where
     StateApi: HasStateApi,
 {
     /// Ensures a value is in the entry by inserting the default value if empty.
+    #[allow(clippy::unwrap_or_default)]
```
Comment (rasmus-kirk) [inline concordium-std/src/impls.rs:768]: Clippy, unhelpfully, suggested to use `self.or_default()` instead...

**Review by abizjak — APPROVED** [review]

**Pull request merged**